lib/spack/spack/new_installer.py

```diff
@@ -0,0 +1,1194 @@
+
+    def __init__(self, control: Connection, parent: Connection) -> None:
+        self.control = control
+        self.parent = parent
+        dev_null_fd = os.open(os.devnull, os.O_WRONLY)
+        #: The file descriptor of the log file (initially /dev/null)
+        self.log_fd = os.dup(dev_null_fd)
+        os.close(dev_null_fd)
+        r, w = os.pipe()
+        self.tee_thread = threading.Thread(
+            target=tee,
+            args=(self.control.fileno(), r, self.log_fd, self.parent.fileno()),
+            daemon=True,
+        )
+        self.tee_thread.start()
+        os.dup2(w, sys.stdout.fileno())
+        os.dup2(w, sys.stderr.fileno())
+        os.close(w)
+
+    def set_output_file(self, path: str) -> None:
+        """Redirect output to the specified log file."""
+        log_fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
+        os.dup2(log_fd, self.log_fd)
+        os.close(log_fd)
+
+    def close(self) -> None:
+        # Closing stdout and stderr should close the last reference to the write end of the pipe,
+        # causing the tee thread to wake up, flush the last data, and exit.
+        os.close(sys.stdout.fileno())
+        os.close(sys.stderr.fileno())
+        self.tee_thread.join()
+        # Only then close the other fds.
+        self.control.close()
+        self.parent.close()
+        os.close(self.log_fd)
+
+
+def install_from_buildcache(
+    mirrors: List[spack.url_buildcache.MirrorURLAndVersion],
+    spec: spack.spec.Spec,
+    unsigned: Optional[bool],
+    state_stream: io.TextIOWrapper,
+) -> bool:
+    send_state("fetching from build cache", state_stream)
+    tarball_stage = spack.binary_distribution.download_tarball(spec.build_spec, unsigned, mirrors)
+
+    if tarball_stage is None:
+        return False
+
+    send_state("relocating", state_stream)
+    spack.binary_distribution.extract_tarball(spec, tarball_stage, force=False)
+
+    if spec.spliced:  # overwrite old metadata with new
+        spack.store.STORE.layout.write_spec(spec, spack.store.STORE.layout.spec_file_path(spec))
+
+    # now a block of curious things follow that should be fixed.
+    pkg = spec.package
+    if hasattr(pkg, "_post_buildcache_install_hook"):
+        pkg._post_buildcache_install_hook()
+    pkg.installed_from_binary_cache = True
+
+    return True
+
+
+def worker_function(
+    spec: spack.spec.Spec,
+    explicit: bool,
+    mirrors: List[spack.url_buildcache.MirrorURLAndVersion],
+    unsigned: Optional[bool],
+    state: Connection,  # write end of state pipe
+    parent: Connection,  # write end of output pipe to parent
+    echo_control: Connection,  # read end of control pipe from parent
+    js1: Optional[Connection],  # ensure that old style jobserver pipes are inherited
+    js2: Optional[Connection],  # ensure that old style jobserver pipes are inherited
+    store: spack.store.Store,
+    config: spack.config.Configuration,
+):
```

Comment:
This needs a docstring to describe what it is/does.

Reply:
Done